src/app/state/operations/tokens.ts

```diff
@@ -0,0 +1,63 @@
+import { Dispatch } from 'react'
+import Eth from 'ethjs'
+import erc720 from './erc720.json'
+import { getAccounts, getNetwork } from '../../../ethrpc'
+import { addToken, addTokenData } from '../reducers/tokens'
+import { getTokens } from '../../../config/getConfig'
+import { saveToLocalStorage, getValueFromLocalStorage } from '../../../features/localStorage'
+
+export const getTokenList = (provider: any) => (dispatch: Dispatch<any>) =>
+  getNetwork(provider)
+    .then((chainId: string) => {
+      const tokens = getTokens(parseInt(chainId))
+      const localTokens = getValueFromLocalStorage(chainId, 'TOKEN')
+      const all = tokens.concat(localTokens)
+      Array.isArray(all) && all.map((address: string) => {
+        dispatch(getToken(provider, address))
+      })
+    })
+
+export const getToken = (provider: any, address: string, onComplete?: any) => (dispatch: Dispatch<any>) => {
+  const eth = new Eth(provider)
+  const token = eth.contract(erc720).at(address)
+
+  eth.getCode(address)
+    .then((result: string) => {
+      if (result === '0x0') {
+        return onComplete ? onComplete(new Error('Not a contract')) : null
+      }
+
+      dispatch(addToken({ address }))
+
+      token.symbol().then((symbol: string[]) => symbol[0])
+        .then((symbol: string) => dispatch(addTokenData({ data: { address, symbol } }))).catch()
+
+      token.name().then((name: string) => name[0])
+        .then((name: string) => dispatch(addTokenData({ data: { address, name } }))).catch()
+
+      getAccounts(provider).then((accounts: string[]) => {
+        token.decimals().then((decimals: any) => {
+          token.balanceOf(accounts[0])
+            .then((balance: any) => {
+              dispatch(addTokenData(
+                { data: { address, balance: balance[0].toString() / Math.pow(10, decimals[0].toNumber()) } }
+              ))
+              onComplete && onComplete(true)
+            })
+        })
+      })
+    })
+}
+
+export const addCustomToken = (provider: any, address: string) => (dispatch: Dispatch<any>) =>
+  new Promise((resolve, reject) => {
```

Comment:
This Promise seems to be unnecessary. It could be handled with the s

Reply:
the 's'?

The promise is returned to the component to let it know to stop 'loading' and either return the error or close the modal

Reply:
Sorry, I want to say _thens_

Reply:
dispatch() does not return a promise.